wrangler-core/src/main/java/io/cdap/wrangler/utils/DirectiveOutputSchemaGenerator.java

```diff
@@ -0,0 +1,111 @@
+ * the License.
+ */
+
+package io.cdap.wrangler.utils;
+
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.wrangler.api.Directive;
+import io.cdap.wrangler.api.Pair;
+import io.cdap.wrangler.api.Row;
+
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * This class can be used to generate the output schema for the output data of a directive. It maintains a map of
+ * output fields present across all output rows after applying a directive. This map is used to generate the schema
+ * if the directive does not return a custom output schema.
+ */
+public class DirectiveOutputSchemaGenerator {
+  private final SchemaConverter schemaGenerator;
+  private final Map<String, Object> outputFieldMap;
+  private final Directive directive;
+
+  public DirectiveOutputSchemaGenerator(Directive directive, SchemaConverter schemaGenerator) {
+    this.directive = directive;
+    this.schemaGenerator = schemaGenerator;
+    outputFieldMap = new LinkedHashMap<>();
+  }
+
+  /**
+   * Method to add new fields from the given output to the map of fieldName --> value maintained for schema generation.
+   * A value is added to the map only if it is absent (or) if the existing value is null and given value is non-null
+   * @param output list of output {@link Row}s after applying directive.
+   */
+  public void addNewOutputFields(List<Row> output) {
+    for (Row row : output) {
+      for (Pair<String, Object> field : row.getFields()) {
+        String fieldName = field.getFirst();
+        Object fieldValue = field.getSecond();
+        if (outputFieldMap.containsKey(fieldName)) {
+          // If existing value is null, override with this non-null value
+          if (fieldValue != null && outputFieldMap.get(fieldName) == null) {
+            outputFieldMap.put(fieldName, fieldValue);
+          }
+        } else {
+          outputFieldMap.putIfAbsent(fieldName, fieldValue);
+        }
+      }
+    }
+  }
+
+  /**
+   * Method to get the output schema of the directive. Returns a generated schema based on maintained map of fields
+   * only if directive does not return a custom output schema.
+   * @param inputSchema input {@link Schema} of the data before applying the directive
+   * @return {@link Schema} corresponding to the output data
+   */
+  public Schema getDirectiveOutputSchema(Schema inputSchema) throws RecordConvertorException {
+    Schema directiveOutputSchema = directive.getOutputSchema(inputSchema);
+    return directiveOutputSchema != null ? directiveOutputSchema : generateDirectiveOutputSchema(inputSchema);
+  }
+
+  // Given the schema from previous step and output of current directive, generates the directive output schema.
+  private Schema generateDirectiveOutputSchema(Schema inputSchema)
+    throws RecordConvertorException {
+    List<Schema.Field> outputFields = new LinkedList<>();
```

Comment:
nit: In 90% of cases LinkedList is slower than ArrayList. Use ArrayList unless you have very specific reasons to use LinkedList